### PR TITLE
Collect heap dumps on Go worker panic for Dataflow

### LINKED ISSUE #21797: [Feature Request]: [Go SDK] Set GOTRACEBACK=crash and recover Go worker heapdumps
### What would you like to happen?

Ideally the worker binary is resilient to crashes with bundles being independant. Some crashes are unavoidable, like OOMs. In such events, it's useful to have a heapdump or core dump to examine to see what could be fixed or improved.

The Go SDK container boot launcher should set `GOTRACEBACK=crash` which will cause unix based systems to write out a core dump on unexpected exits (such as OOMs and Similar). Further, like with java, a restarting process should try to find any dump written, and move it to a target location (like the job's designated temp folder).

See https://pkg.go.dev/runtime and https://pkg.go.dev/runtime/debug#SetTraceback for some details.

Complications:
1. Collecting the core dump.
Based on the test for such crash dumps, it looks like it may by default get written to StdErr out. https://cs.opensource.google/go/go/+/refs/tags/go1.18.3:src/runtime/crash_unix_test.go;l=163
We can't simply write all std err to a file, because for long running tasks that would become extremely large from other debug printouts.
This should be validated, and default std.err logging changed in this case.

But this could be a misreading. Traceback information is printed (stack traces etc), but the heap dump might still go elsewhere. This needs to be determined, and in particular if docker containers are used, whether the file itself is accessible to the next container started by the VM.

2. Moving the file
Since the pipeline binary would have the facility to write to the job's temp directory, such copying code would need to live in the harness, (likely in harness/init.go), so it can write the file if available.

Alternatively, there's possibly some configuration that could be done for the runner to manage this instead, but this isn't clear offhand.

Java's approach is contained in [MemoryMonitor](https://github.com/apache/beam/blob/4ffeae4d2b800f2df36d2ea2eab549f2204d5691/sdks/java/harness/src/main/java/org/apache/beam/fn/harness/status/MemoryMonitor.java)

### Issue Priority

Priority: 2

### Issue Component

Component: sdk-go

### PR DESCRIPTION
**Pull request by damccorm**
This collects a heap dump and uploads it to the temp_directory on Dataflow any time there is a worker panic before we exit. It departs from what #21797 recommends a little bit in that it doesn't use `GOTRACEBACK=crash` and instead relies on `debug.WriteHeapDump`. This is mostly because the GOTRACEBACK approach doesn't collect the dump until the top level Go routine exits with a panic - this currently isn't guaranteed to happen (and won't by default because we recover from panics and call os.exit), but if it did it would propogate up through boot.go. At that point, the VM is not guaranteed to be persisted (in fact, I think it will always be torn down), and this generally just breaks how we handle exiting. The approach of gathering the dump ourselves is also more in line with how Java handles this anyways - https://github.com/apache/beam/blob/4ffeae4d2b800f2df36d2ea2eab549f2204d5691/sdks/java/harness/src/main/java/org/apache/beam/fn/harness/status/MemoryMonitor.java#L589

To test this, I inserted a panic into the harness code and verified that a heap dump was correctly uploaded to the temp directory:

- pipeline - https://pantheon.corp.google.com/dataflow/jobs/us-central1/2022-07-08_07_34_58-5641449712707134536;bottomTab=WORKER_LOGS;logsSeverity=INFO?project=google.com:clouddfe&pageState=(%22dfTime%22:(%22l%22:%22dfJobMaxTime%22))
- dumps from multiple pipeline runs - https://pantheon.corp.google.com/storage/browser/clouddfe-test/dannymccormick-tmp/heapDumps?pageState=(%22StorageObjectListTable%22:(%22f%22:%22%255B%255D%22))&project=google.com:clouddfe&prefix=&forceOnObjectsSortingFiltering=false)

Fixes #21797

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [x] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Comment by lostluck** [comment]
I was under the impression that there were panics we couldn't use "recover" from, such as
OOM and concurrent map writes+reads.  Otherwise, this would be something else we need to bolster our "callNoPanic" path with since an OOM'd worker isn't likely to be able to recover.

See https://stackoverflow.com/questions/57486620/are-all-runtime-errors-recoverable-in-go for some examples on how to trigger them.

It does seem like we could probably have an actual integration test for this too that runs on Dataflow, pipeline that will start, and then OOM, and then we could observe the heap dump was written to the given GCS folder.

File: sdks/go/pkg/beam/core/runtime/harness/init/init.go [hunk 1]
```diff
@@ -88,10 +90,24 @@ func hook() {
...
+	ctx = grpcx.WriteWorkerID(ctx, *id)
 	defer func() {
 		if r := recover(); r != nil {
 			fmt.Fprintf(os.Stderr, "Worker panic: %v\n", r)
 			debug.PrintStack()
+			log.Warnf(ctx, "Worker panic detected: %v", r)
+
+			if tempLocation := beam.PipelineOptions.Get("temp_location"); tempLocation != "" {
+				log.Warnf(ctx, "Capturing heap dump")
+				harness.UploadHeapDump(ctx, fmt.Sprintf("%v/heapDumps/dump-%d", strings.TrimSuffix(tempLocation, "/"), time.Now().Unix()))
```
Comment (lostluck) [inline sdks/go/pkg/beam/core/runtime/harness/init/init.go:108]: Please also use the worker-id `*id` and not just the time.

File: sdks/go/pkg/beam/runners/dataflow/dataflow.go [hunk 2]
```diff
@@ -117,18 +117,19 @@ var flagFilter = map[string]bool{
...
-	"dataflow_worker_jar":            true,
-	"worker_region":                  true,
-	"worker_zone":                    true,
-	"teardown_policy":                true,
-	"cpu_profiling":                  true,
-	"session_recording":              true,
-	"update":                         true,
-	"transform_name_mapping":         true,
+	// Don't filter temp_location since we need this included in PipelineOptions to correctly upload heap dumps.
+	// "temp_location":                  true,
```
Comment (lostluck) [inline sdks/go/pkg/beam/runners/dataflow/dataflow.go:121]: I'd remove this line entirely, but still have a comment about it somewhere.

**Review by lostluck — CHANGES_REQUESTED** [review]

**Comment by lostluck** [comment]
Also, it's very easy to subvert a recover:
```
go func(){  panic("escape recover") }
```

**Comment by damccorm** [comment]
> See https://stackoverflow.com/questions/57486620/are-all-runtime-errors-recoverable-in-go for some examples on how to trigger them.

Ah, that's helpful context. I'll need to play around with how this actually behaves, but I agree this isn't the right approach. I'll give this another pass

**Pull request closed without merging**